rcgen/docs/0.12-to-0.13.md

```diff
@@ -0,0 +1,90 @@
+# Rcgen 0.12 to 0.13 Migration Guide
+
+This document is a meant to be a helpful guide for some of the API changes made
+between rcgen 0.12 and 0.13. For information on other changes in 0.13 see
+[rcgen/CHANGELOG.md].
+
+## Key Pairs
+
+* Previously it was possible to have certificate generation automatically create
+  a subject `KeyPair` for you by leaving the `key_pair` field of
+  `CertificateParams` empty, and retrieving the generated `KeyPair` from
+  a `Certificate` created with the `CertificateParams` by calling
+  `Certificate::get_key_pair()`.
+
+  To offer more consistency and to keep the `CertificateParams` and `Certificate`
+  types from holding private key data, the new API requires you handle `KeyPair`
+  creation yourself. See `CertifiedKey`, `KeyPair::generate()`,
+  `KeyPair::generate_for()` and `KeyPair::generate_rsa_for()` for more information.
+
+* Serializing a `Certificate`'s `KeyPair` to DER or PEM was previously done by
+  calling `Certificate::serialize_private_key_der()` or
+  `Certificate::serialize_private_key_pem()`. This is now handled by calling
+  `KeyPair::serialize_der()` or `KeyPair::serialize_pem()`.
+
+## Certificates
+
+* For quick-and-easy self-signed certificate issuance,
+  `generate_simple_self_signed` now returns a `CertifiedKey` in the success case
+  instead of a `Certificate`. The self-signed `Certificate` can be accessed in
+  the `cert` field of `CertifiedKey`, and the generated subject key pair in
+  `key_pair`.
+
+* Customized self-signed certificate issuance was previously done by
+  constructing `CertificateParams` and calling `Certificate::from_params()` to
+  create a `Certificate`. This is now done by calling
+  `CertificateParams::self_signed()`, providing a subject `KeyPair` of your
+  choosing.
+
+* Customized certificate issuance signed by an issuer was previously done by
+  constructing `CertificateParams`, calling `Certificate::from_params()` and
+  then choosing the issuer at serialization time. This is now done ahead of
+  serialization by calling `CertificateParams::signed_by()` and providing
+  a subject `KeyPair` as well as an issuer `Certificate` and `KeyPair`.
+
+* Previously certificate serialization was done by calling
+  `Certificate::serialize_der()`, `Certificate::serialize_pem()`,
+  `Certificate::serialize_der_with_signer()` or
+  `Certificate::serialize_pem_with_signer()`. Each time a serialization fn was
+  called a new certificate was issued, leading to confusion when it was desired
+  to serialize the same certificate in two formats. In the new API issuance is
+  handled by `CertificateParams` fns and a `Certificate` is stable. You can
```

Comment:
> or explicitly say "and the generated Certificate does not change"?

I went with a small tweak on this language saying "the generated `Certificate` will not change when serialized". Good suggestion, thanks.

Reply:
It kind of still feels like a silly thing to say because the `der()` method doesn't even really do serialization and the new API just intrinsically makes it impossible for the previous pitfall to happen... not really sure how best to explain this.

Reply:
> It kind of still feels like a silly thing to say because the der() method doesn't even really do serialization and the new API just intrinsically makes it impossible for the previous pitfall to happen.

I think the reason it feels silly is because the old API was surprising and the new behaviour seems intuitively like what most folks would expect. 

I still think the text is better explicitly saying both what the old API did and what the new API doesn't do. I suspect many readers won't have realized what the old API was doing until this moment and might want to check their own usage for bugs.

> the der() method doesn't even really do serialization

No argument there. I changed the last sentence to say "You can serialize it to PEM by calling `Certificate::pem()`, or access the DER encoding by calling `Certificate::der()`.